<!DOCTYPE html><html lang="zh-cmn-Hans" prefix="og: http://ogp.me/ns#" class="han-init"><head><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" /><title>宏基因组分析流程 &mdash; Wanjin Hu (胡万金)</title><link rel="stylesheet" href="http://localhost:4000/assets/vendor/primer-css/css/primer.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/components/collection.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/components/repo-card.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/sections/repo-list.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/components/boxed-group.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/globals/common.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/globals/responsive.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/posts/index.css"><link rel="stylesheet" href="http://localhost:4000/assets/vendor/octicons/octicons/octicons.css"><link rel="stylesheet" href="https://mazhuang.org/rouge-themes/dist/github.css"><link rel="canonical" href="http://localhost:4000/2023/08/10/pipeline-metagenomic-analysis/"><link rel="alternate" type="application/atom+xml" title="Wanjin Hu (胡万金)" href="http://localhost:4000/feed.xml"><link rel="shortcut icon" href="http://localhost:4000/favicon.ico"><meta property="og:title" content="宏基因组分析流程"><meta name="keywords" content="Metagenomic, Microbiome"><meta name="og:keywords" content="Metagenomic, Microbiome"><meta name="description" content="提供一个宏基因组分析流程。"><meta name="og:description" content="提供一个宏基因组分析流程。"><meta property="og:url" content="http://localhost:4000/2023/08/10/pipeline-metagenomic-analysis/"><meta property="og:site_name" content="Wanjin Hu (胡万金)"><meta property="og:type" content="article"><meta property="og:locale" content="zh_CN" /><meta property="article:published_time" content="2023-08-10"> <script src="http://localhost:4000/assets/vendor/jquery/dist/jquery.min.js"></script> <script src="http://localhost:4000/assets/js/main.js"></script></head><body class="" data-mz=""><header class="site-header"><div class="container"><h1><a href="http://localhost:4000/" title="Wanjin Hu (胡万金)"><span class="octicon octicon-mark-github"></span> Wanjin Hu (胡万金)</a></h1><button class="collapsed mobile-visible" type="button" onclick="toggleMenu();"> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button><nav class="site-header-nav" role="navigation"> <a href="http://localhost:4000/" class="site-header-nav-item" target="" title="首页">首页</a> <a href="http://localhost:4000/categories/" class="site-header-nav-item" target="" title="分类">分类</a> <a href="http://localhost:4000/archives/" class="mobile-hidden site-header-nav-item" target="" title="归档">归档</a> <a href="http://localhost:4000/open-source/" class="mobile-hidden site-header-nav-item" target="" title="开源">开源</a> <a href="http://localhost:4000/fragments/" class="site-header-nav-item" target="" title="片段">片段</a> <a href="http://localhost:4000/wiki/" class="site-header-nav-item" target="" title="维基">维基</a> <a href="http://localhost:4000/links/" class="mobile-hidden site-header-nav-item" target="" title="链接">链接</a> <a href="http://localhost:4000/about/" class="site-header-nav-item" target="" title="关于">关于</a> <a class="mobile-hidden" href="http://localhost:4000/feed.xml"><span class="octicon octicon-rss" style="color:orange;"></span></a></nav></div></header><section class="collection-head small geopattern" data-pattern-id="宏基因组分析流程"><div class="container"><div class="columns"><div class="column three-fourths"><div class="collection-title"><h1 class="collection-header">宏基因组分析流程</h1><div class="collection-info"> <span class="meta-info"> <span class="octicon octicon-calendar"></span> 2023/08/10 </span> <span class="meta-info"> <span class="octicon octicon-file-directory"></span> <a href="http://localhost:4000/categories/#Metagenomic" title="Metagenomic">Metagenomic</a> </span> <span class="meta-info"> <span class="octicon octicon-clock"></span> 共 7201 字，约 21 分钟 </span></div></div></div><div class="column one-fourth mobile-hidden"><div class="collection-title"></div></div></div></div></section><section class="container content"><div class="columns"><div class="column three-fourths" ><article class="article-content markdown-body"><p>提供一个宏基因组分析流程。</p><p>Here is a metagenomic sequence data analysis pipeline, nothing different with other pipelines. But if you want to get to know the metagenomic analysis pipeline step by step, maybe you can get some details from this repository. And it is suitable for the beginners i think.</p><p>Scripts and test files you can find here: <a href="https://github.com/wanjinhu/Metagenomic-Analysis-Pipeline">Metagenomic-Analysis-Pipeline</a></p><h2 id="pipeline-overview-">Pipeline overview 🐫</h2><ul><li>Raw sequence quality trim</li><li>Host reference sequence remove</li><li>Metaphlan for composition of microbial communities</li><li>Sequence assembly</li><li>Gene prediction</li><li>Remove redundancy gene and build non-redundant geneset</li><li>Function annotation using emapper</li><li>Organize function results table</li></ul><h2 id="quick-start-">Quick start 🦏</h2><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nv">$python</span> pipe_metagenome.py <span class="nt">-h</span>
usage: 
<span class="o">=================================================================</span>
python pipe_metagenome.py
	<span class="nt">--fastq_list</span> fq.list
	<span class="nt">--output_dir</span> result
	<span class="nt">--ref</span> ref_bowtie2_index
ref_bowtie2_index:
canis: /root/database/Canis_GCF_000002285.5/Canis_GCF_000002285_5
human: /root/database/hg38_GCF_000001405.40/GCF_000001405.40/hg38
<span class="o">=================================================================</span>

Pipeline of metagenome

optional arguments:
  <span class="nt">-h</span>, <span class="nt">--help</span>            show this <span class="nb">help </span>message and <span class="nb">exit</span>
  <span class="nt">-l</span> FQLIST, <span class="nt">--fastq_list</span> FQLIST
                        raw fq list
  <span class="nt">-o</span> OUTDIR, <span class="nt">--output_dir</span> OUTDIR
                        result output
  <span class="nt">-r</span> REF, <span class="nt">--ref</span> REF     ref genome bowtie2 index
</code></pre></div></div><p>What you need to do is to provide two input files:</p><ul><li>–fastq_list # sample - fq_R1 - fq_R2 list, format like <code class="language-plaintext highlighter-rouge">fq.list</code></li><li>–ref # host reference genome bowtie2 index</li></ul><p>And set output dir <code class="language-plaintext highlighter-rouge">--output_dir</code>, all of output results would be included.</p><h2 id="output-files-explanation-">Output files explanation 🐊</h2><h3 id="output-files-tree-not-show-all-files">Output files tree (not show all files)</h3><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>├── 00-result                           <span class="c"># most important results in this fold</span>
│   ├── 00_merged_abundance_table.txt   <span class="c"># composition of microbial communities</span>
│   ├── 01_metaphlan_phylum.txt         <span class="c"># communities in phylum level</span>
│   ├── 02_metaphlan_class.txt          <span class="c"># communities in class level</span>
│   ├── 03_metaphlan_order.txt          <span class="c"># communities in order level</span>
│   ├── 04_metaphlan_family.txt         <span class="c"># communities in family level</span>
│   ├── 05_metaphlan_genus.txt          <span class="c"># communities in genus level</span>
│   ├── 06_metaphlan_species.txt        <span class="c"># communities in species level</span>
│   ├── KO_samples.xls                  <span class="c"># KEGG KO gene composition table</span>
│   └── pathway_samples.xls             <span class="c"># KEGG pathway composition table</span>
├── 01-fastp_trim
├── 02-ref_remove
├── 03-metaphlan
├── 04-megahit
├── 05-prodigal
├── 06-cdhit
├── 07-emapper
├── 08-sam_count
├── 09-emapper_kegg
</code></pre></div></div><h3 id="output-important-result-files-explanation">Output important result files explanation</h3><h4 id="metaphlan_diff-levelstxt">metaphlan_diff-levels.txt</h4><ul><li>column 1 : communities information</li><li>column 2 ~ : communities abundance percent of samples<div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>clade_name	C1
k__Bacteria<span class="p">;</span>p__Firmicutes	86.51132
k__Bacteria<span class="p">;</span>p__Actinobacteria	6.52203
k__Bacteria<span class="p">;</span>p__Bacteroidetes	4.24729
k__Bacteria<span class="p">;</span>p__Proteobacteria	1.96422
k__Bacteria<span class="p">;</span>p__Fusobacteria	0.75514
k__Bacteria<span class="p">;</span>p__Tenericutes	0.0
k__Bacteria<span class="p">;</span>p__Spirochaetes	0.0
k__Bacteria<span class="p">;</span>p__Verrucomicrobia	0.0
...
</code></pre></div></div><h4 id="ko_samplesxls">KO_samples.xls</h4></li><li>column 1 : KO gene name</li><li>column 2 : KO gene description</li><li>column 3 : KO gene id</li><li>column 4 ~ : KO gene number of samples<div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>KO_name	KO_des	KO	C1
<span class="s2">"E1.1.1.1, adh"</span>	alcohol dehydrogenase <span class="o">[</span>EC:1.1.1.1]	K00001	2817.0
<span class="s2">"AKR1A1, adh"</span>	alcohol dehydrogenase <span class="o">(</span>NADP+<span class="o">)</span> <span class="o">[</span>EC:1.1.1.2]	K00002	254.0
hom	homoserine dehydrogenase <span class="o">[</span>EC:1.1.1.3]	K00003	2890.0
<span class="s2">"BDH, butB"</span>	<span class="s2">"(R,R)-butanediol dehydrogenase / meso-butanediol dehydrogenase / diacetyl reductase [EC:1.1.1.4 1.1.1.- 1.1.1.303]"</span>	K00004	20.0
...
</code></pre></div></div><h4 id="pathway_samplesxls">pathway_samples.xls</h4></li><li>column 1 : pathway level 1</li><li>column 2 : pathway level 2</li><li>column 3 : pathway level 3</li><li>column 4 : pathway id</li><li>column 5 ~ : pathway gene number of samples<div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>level1	level2	level3	pathway	C1
Metabolism	Carbohydrate metabolism	Glycolysis / Gluconeogenesis	ko00010	91005.0
Metabolism	Carbohydrate metabolism	Citrate cycle <span class="o">(</span>TCA cycle<span class="o">)</span>	ko00020	31442.0
Metabolism	Carbohydrate metabolism	Pentose phosphate pathway	ko00030	53905.0
Metabolism	Carbohydrate metabolism	Pentose and glucuronate interconversions	ko00040	21334.0
...
</code></pre></div></div></li></ul><h2 id="step-by-step-">Step by step 🦥</h2><h3 id="step1-raw-sequence-quality-trim-using-fastp">Step1 Raw sequence quality trim using fastp</h3><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>fastp <span class="nt">-i</span> sample_1.fastq.gz <span class="se">\</span>
      <span class="nt">-o</span> sample_clean.1.fastq.gz <span class="se">\</span>
      <span class="nt">-I</span> sample_2.fastq.gz <span class="se">\</span>
      <span class="nt">-O</span> sample_clean.2.fastq.gz <span class="se">\</span>
      <span class="nt">-w</span> 8 <span class="nt">-h</span> sample.html <span class="nt">-j</span> sample.json
</code></pre></div></div><h3 id="step2-host-reference-sequence-remove">Step2 Host reference sequence remove</h3><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bowtie2 <span class="nt">-x</span> ref_bowtie2_index <span class="nt">-1</span> sample_clean.1.fastq.gz <span class="nt">-2</span> sample_clean.2.fastq.gz <span class="nt">-S</span> sample.sam  2&gt;sample.mapping.log
samtools fastq -@ 8 <span class="nt">-f</span> 4 sample.sam <span class="nt">-1</span> sample.unmap.1.fastq.gz <span class="nt">-2</span> sample.unmap.2.fastq.gz <span class="nt">-s</span> sample.unmap.single.fastq.gz
</code></pre></div></div><h3 id="step3-metaphlan-for-composition-of-microbial-communities">Step3 Metaphlan for composition of microbial communities</h3><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>zcat sample.unmap.1.fastq.gz sample.unmap.2.fastq.gz|metaphlan <span class="nt">--input_type</span> fastq <span class="nt">--bowtie2out</span> sample_bowtie2.bz2 <span class="nt">--output_file</span> sample_metaphlan.tsv <span class="nt">--nproc</span> 8
<span class="c"># when you install metaphlan in the system, you will get script 'merge_metaphlan_tables.py', that's for merge different samples metaphlan result in one file, like: </span>
merge_metaphlan_tables.py <span class="k">*</span>.tsv <span class="o">&gt;</span> 00_merged_abundance_table.txt
<span class="nb">grep</span> <span class="nt">-E</span> <span class="s1">'(p__)|(clade_name)'</span> 00_merged_abundance_table.txt |grep <span class="nt">-v</span> <span class="s1">'c__'</span>|sed <span class="s1">'s/|/;/g'</span> <span class="o">&gt;</span> 01_metaphlan_phylum.txt
<span class="nb">grep</span> <span class="nt">-E</span> <span class="s1">'(c__)|(clade_name)'</span> 00_merged_abundance_table.txt |grep <span class="nt">-v</span> <span class="s1">'o__'</span>|sed <span class="s1">'s/|/;/g'</span> <span class="o">&gt;</span> 02_metaphlan_class.txt
<span class="nb">grep</span> <span class="nt">-E</span> <span class="s1">'(o__)|(clade_name)'</span> 00_merged_abundance_table.txt |grep <span class="nt">-v</span> <span class="s1">'f__'</span>|sed <span class="s1">'s/|/;/g'</span> <span class="o">&gt;</span> 03_metaphlan_order.txt
<span class="nb">grep</span> <span class="nt">-E</span> <span class="s1">'(f__)|(clade_name)'</span> 00_merged_abundance_table.txt |grep <span class="nt">-v</span> <span class="s1">'g__'</span>|sed <span class="s1">'s/|/;/g'</span> <span class="o">&gt;</span> 04_metaphlan_family.txt
<span class="nb">grep</span> <span class="nt">-E</span> <span class="s1">'(g__)|(clade_name)'</span> 00_merged_abundance_table.txt |grep <span class="nt">-v</span> <span class="s1">'s__'</span>|sed <span class="s1">'s/|/;/g'</span> <span class="o">&gt;</span> 05_metaphlan_genus.txt
<span class="nb">grep</span> <span class="nt">-E</span> <span class="s1">'(s__)|(clade_name)'</span> 00_merged_abundance_table.txt |grep <span class="nt">-v</span> <span class="s1">'t__'</span>|sed <span class="s1">'s/|/;/g'</span> <span class="o">&gt;</span> 06_metaphlan_species.txt
</code></pre></div></div><h3 id="step4-sequence-assembly-and-trim-contigs-which-length--500bp">Step4 Sequence assembly and trim contigs which length &lt; 500bp</h3><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>megahit <span class="nt">-1</span> sample.unmap.1.fastq.gz <span class="nt">-2</span> sample.unmap.2.fastq.gz <span class="nt">-o</span> sample_megahit <span class="nt">--out-prefix</span> sample <span class="nt">-t</span> 8
seqkit <span class="nb">seq</span> <span class="nt">-m</span> 500 sample_megahit/sample.contigs.fa <span class="nt">--remove-gaps</span> <span class="o">&gt;</span> sample.contigs_500.fa
<span class="nb">sed</span> <span class="nt">-i</span> <span class="s1">'s/&gt;/&gt;sample_/g'</span> sample.contigs_500.fa
</code></pre></div></div><h3 id="step5-gene-prediction-using-prodigal">Step5 Gene prediction using prodigal</h3><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>prodigal <span class="nt">-p</span> meta <span class="nt">-a</span> sample_prot.faa <span class="nt">-m</span> <span class="nt">-d</span> sample_nucl.fna <span class="nt">-o</span> sample_genes.gff <span class="nt">-f</span> gff <span class="nt">-s</span> sample.stat <span class="nt">-i</span> sample.contigs_500.fa
</code></pre></div></div><h3 id="step6-remove-redundancy-gene-and-build-non-redundant-geneset">Step6 Remove redundancy gene and build non-redundant geneset</h3><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="nb">cat </span>sample1_prot.faa sample2_prot.faa ... <span class="o">&gt;</span>  prot.faa
<span class="nb">cat </span>sample1_nucl.fna sample2_nucl.fna ... <span class="o">&gt;</span>  nucl.fna
cd-hit <span class="nt">-i</span> prot.faa <span class="nt">-o</span> prot_nonerude.faa <span class="nt">-c</span> 0.95 <span class="nt">-T</span> 8 <span class="nt">-n</span> 5 <span class="nt">-d</span> 0 <span class="nt">-aS</span> 0.9 <span class="nt">-g</span> 1 <span class="nt">-sc</span> 1 <span class="nt">-sf</span> 1 <span class="nt">-M</span> 0
<span class="nb">grep</span> <span class="s1">'&gt;'</span> prot_nonerude.faa|awk <span class="nt">-F</span> <span class="s1">' '</span> <span class="s1">'{print $1}'</span>|sed <span class="s1">'s/&gt;//g'</span> <span class="o">&gt;</span> prot_nonerude.list
seqtk subseq nucl.fna prot_nonerude.list <span class="o">&gt;</span> nucl_nonerude.fna
bwa index nucl_nonerude.fna <span class="nt">-p</span> geneset_bwa
bioawk <span class="nt">-c</span> fastx <span class="s1">'{print $name, length($seq)}'</span> nucl_nonerude.fna <span class="o">&gt;</span> geneset_length.txt
</code></pre></div></div><h3 id="step7-function-annotation-using-emapper">Step7 Function annotation using emapper</h3><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># when you install emapper in the system, you will get script 'emapper.py'</span>
emapper.py <span class="nt">-i</span> prot_nonerude.faa <span class="nt">-o</span> eggnog <span class="nt">--cpu</span> 0 <span class="nt">--usemem</span>
<span class="nb">cut</span> <span class="nt">-f1</span>,12 eggnog.emapper.annotations|grep <span class="nt">-v</span> <span class="s2">"^#"</span>|sed <span class="s1">'s/ko://g'</span>|sed <span class="s1">'1i gene\tko'</span>|grep <span class="nt">-v</span> <span class="s2">"-"</span> <span class="o">&gt;</span> KEGG_KO.txt
<span class="nb">cut</span> <span class="nt">-f1</span>,13 eggnog.emapper.annotations|grep <span class="nt">-v</span> <span class="s2">"^#"</span>|sed <span class="s1">'1i gene\tpathway'</span>|grep <span class="nt">-v</span> <span class="s2">"-"</span> <span class="o">&gt;</span> KEGG_PATHWAY.txt
</code></pre></div></div><h3 id="step8-gene-num-count">Step8 Gene num count</h3><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bwa mem <span class="nt">-t</span> 4 geneset_bwa sample.unmap.1.fastq.gz sample.unmap.2.fastq.gz | samtools view <span class="nt">-bS</span> - | samtools <span class="nb">sort</span> - <span class="o">&gt;</span> sample_mapping_geneset.bam
samtools view <span class="nt">-F</span> 4 <span class="nt">-F</span> 256 <span class="nt">-F</span> 2048 sample_mapping_geneset.bam|awk <span class="s1">'{if($3!="*") print $3}'</span>|sort| <span class="nb">uniq</span> <span class="nt">-c</span>|awk <span class="s1">'BEGIN {FS=" ";OFS=","} {print $2,$1}'</span> | <span class="nb">awk</span> <span class="s1">'BEGIN {FS=",";OFS=","} {if ($2 &gt; 1) print $1"\t"$2; else print $1"\t0"}'</span>|sed <span class="s1">'1i gene\tsample'</span> <span class="o">&gt;</span> sample.count
</code></pre></div></div><h3 id="step9-organize-function-results-table">Step9 Organize function results table</h3><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># using /kegg/kegg.py to analysis, like:</span>
<span class="nv">$python</span> kegg.py <span class="nt">-h</span>
usage: python kegg.py <span class="nt">-kk</span> KEGG_KO.txt <span class="nt">-kp</span> KEGG_PATHWAY.txt <span class="nt">-mt</span> merged_file.txt <span class="nt">-ok</span> out_KO.xls <span class="nt">-op</span> out_pathway.xls

Merge KO/pathway count table from eggnog result.

optional arguments:
  <span class="nt">-h</span>, <span class="nt">--help</span>            show this <span class="nb">help </span>message and <span class="nb">exit</span>
  <span class="nt">-kk</span> KEGGKO, <span class="nt">--kegg_KO</span> KEGGKO
                        Sample<span class="s1">'s kegg KO information, such as KEGG_KO.txt
  -kp KEGGPATHWAY, --kegg_pathway KEGGPATHWAY
                        Sample'</span>s kegg pathway information, such as
                        KEGG_PATHWAY.txt
  <span class="nt">-mt</span> MERGETABLE, <span class="nt">--merge_table</span> MERGETABLE
                        Sample<span class="s1">'s merged gene count table, such as
                        merged_file.txt
  -ok OUTKO, --out_KO OUTKO
                        Output KO result, such as out_KO.xls
  -op OUTPATHWAY, --out_pathway OUTPATHWAY
                        Output pathway result, such as out_pathway.xls
  -t TMP, --tmp TMP     Tmp files dir
</span></code></pre></div></div><h2 id="contact-">Contact 🐖</h2><p>Wanjin Hu (wanjin.hu@outlook.com)</p><div style="margin-top:2em;padding:0 1.5em;border:1px solid #d3d3d3;background-color:#deebf7"><h3>文档信息</h3><ul><li>本文作者：<a href="http://localhost:4000" target="_blank">Wanjin Hu (胡万金)</a></li><li>本文链接：<a href="http://localhost:4000/2023/08/10/pipeline-metagenomic-analysis/" target="_blank">http://localhost:4000/2023/08/10/pipeline-metagenomic-analysis/</a></li><li>版权声明：自由转载-非商用-非衍生-保持署名（<a href="http://creativecommons.org/licenses/by-nc-nd/3.0/deed.zh" target="_blank">创意共享3.0许可证</a>）</li></ul></div></article><div class="share"></div><div class="comment"> <script src="https://giscus.app/client.js" data-repo="wanjinhu/wanjinhu.github.io" data-repo-id="R_kgDOKJb3gA" data-category="Announcements" data-category-id="DIC_kwDOKJb3gM4CYvQX" data-mapping="title" data-strict="1" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="zh-CN" data-loading="lazy" crossorigin="anonymous" async> </script></div></div><div class="column one-fourth"><h3>Search</h3><div id="site_search"> <input style="width:96%" type="text" id="search_box" placeholder="Search"></div><ul id="search_results" style="font-size:14px;list-style-type:none;padding-top:10px;padding-left:10px;"></ul><script src="http://localhost:4000/assets/js/simple-jekyll-search.min.js"></script> <script type="text/javascript"> SimpleJekyllSearch({ searchInput: document.getElementById('search_box'), resultsContainer: document.getElementById('search_results'), json: 'http://localhost:4000/assets/search_data.json?v=1693363113', searchResultTemplate: '<li><a href="{url}" title="{title}">{title}</a></li>', noResultsText: 'No results found', limit: 10, fuzzy: false, exclude: ['Welcome'] }) </script><h3 class="post-directory-title mobile-hidden">Table of Contents</h3><div id="post-directory-module" class="mobile-hidden"><section class="post-directory"><dl></dl></section></div><script src="http://localhost:4000/assets/js/jquery.toc.js"></script></div></div></section><footer class="container"><div class="site-footer" role="contentinfo"><div class="copyright left mobile-block"> © 2023 <span title="Wanjin Hu (胡万金)">Wanjin Hu (胡万金)</span> <a href="javascript:window.scrollTo(0,0)" class="right mobile-visible">TOP</a></div><ul class="site-footer-links right mobile-hidden"><li> <a href="javascript:window.scrollTo(0,0)" >TOP</a></li></ul><a href="https://github.com/wanjinhu/wanjinhu.github.io" target="_blank" aria-label="view source code"> <span class="mega-octicon octicon-mark-github" title="GitHub"></span> </a><ul class="site-footer-links mobile-hidden"><li> <a href="http://localhost:4000/" title="首页" target="">首页</a></li><li> <a href="http://localhost:4000/categories/" title="分类" target="">分类</a></li><li> <a href="http://localhost:4000/archives/" title="归档" target="">归档</a></li><li> <a href="http://localhost:4000/open-source/" title="开源" target="">开源</a></li><li> <a href="http://localhost:4000/fragments/" title="片段" target="">片段</a></li><li> <a href="http://localhost:4000/wiki/" title="维基" target="">维基</a></li><li> <a href="http://localhost:4000/links/" title="链接" target="">链接</a></li><li> <a href="http://localhost:4000/about/" title="关于" target="">关于</a></li><li><a href="http://localhost:4000/feed.xml"><span class="octicon octicon-rss" style="color:orange;"></span></a></li></ul></div></footer><div class="tools-wrapper"> <a class="gotop" href="#" title="回到顶部"><span class="octicon octicon-arrow-up"></span></a></div><script src="http://localhost:4000/assets/js/geopattern.js"></script> <script> jQuery(document).ready(function($) { $('.geopattern').each(function(){ $(this).geopattern($(this).data('pattern-id')); }); /* hljs.initHighlightingOnLoad(); */ }); </script></body></html>
