<!DOCTYPE html><html lang="zh-cmn-Hans" prefix="og: http://ogp.me/ns#" class="han-init"><head><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" /><title>QIIME2 构建数据库以及常规分析流程 &mdash; Wanjin Hu (胡万金)</title><link rel="stylesheet" href="http://localhost:4000/assets/vendor/primer-css/css/primer.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/components/collection.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/components/repo-card.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/sections/repo-list.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/components/boxed-group.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/globals/common.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/globals/responsive.css"><link rel="stylesheet" href="http://localhost:4000/assets/css/posts/index.css"><link rel="stylesheet" href="http://localhost:4000/assets/vendor/octicons/octicons/octicons.css"><link rel="stylesheet" href="https://mazhuang.org/rouge-themes/dist/github.css"><link rel="canonical" href="http://localhost:4000/2023/10/27/pipeline-qiime2-analysis/"><link rel="alternate" type="application/atom+xml" title="Wanjin Hu (胡万金)" href="http://localhost:4000/feed.xml"><link rel="shortcut icon" href="http://localhost:4000/favicon.ico"><meta property="og:title" content="QIIME2 构建数据库以及常规分析流程"><meta name="keywords" content="QIIME2，测序数据分析"><meta name="og:keywords" content="QIIME2，测序数据分析"><meta name="description" content="QIIME2 分析流程是扩增子分析常用的流程之一，这里提供一个完整的 16S 数据分析流程，包括构建数据库以及常规分析。QIIME2 是各种插件组成的综合体 pipeline，具体的信息可以看官方网站 QIIME2."><meta name="og:description" content="QIIME2 分析流程是扩增子分析常用的流程之一，这里提供一个完整的 16S 数据分析流程，包括构建数据库以及常规分析。QIIME2 是各种插件组成的综合体 pipeline，具体的信息可以看官方网站 QIIME2."><meta property="og:url" content="http://localhost:4000/2023/10/27/pipeline-qiime2-analysis/"><meta property="og:site_name" content="Wanjin Hu (胡万金)"><meta property="og:type" content="article"><meta property="og:locale" content="zh_CN" /><meta property="article:published_time" content="2023-10-27"> <script src="http://localhost:4000/assets/vendor/jquery/dist/jquery.min.js"></script> <script src="http://localhost:4000/assets/js/main.js"></script></head><body class="" data-mz=""><header class="site-header"><div class="container"><h1><a href="http://localhost:4000/" title="Wanjin Hu (胡万金)"><span class="octicon octicon-mark-github"></span> Wanjin Hu (胡万金)</a></h1><button class="collapsed mobile-visible" type="button" onclick="toggleMenu();"> <span class="icon-bar"></span> <span class="icon-bar"></span> <span class="icon-bar"></span> </button><nav class="site-header-nav" role="navigation"> <a href="http://localhost:4000/" class="site-header-nav-item" target="" title="首页">首页</a> <a href="http://localhost:4000/categories/" class="site-header-nav-item" target="" title="分类">分类</a> <a href="http://localhost:4000/archives/" class="mobile-hidden site-header-nav-item" target="" title="归档">归档</a> <a href="http://localhost:4000/open-source/" class="mobile-hidden site-header-nav-item" target="" title="开源">开源</a> <a href="http://localhost:4000/wiki/" class="site-header-nav-item" target="" title="维基">维基</a> <a href="http://localhost:4000/links/" class="mobile-hidden site-header-nav-item" target="" title="链接">链接</a> <a href="http://localhost:4000/about/" class="site-header-nav-item" target="" title="关于">关于</a> <a class="mobile-hidden" href="http://localhost:4000/feed.xml"><span class="octicon octicon-rss" style="color:orange;"></span></a></nav></div></header><section class="collection-head small geopattern" data-pattern-id="QIIME2 构建数据库以及常"><div class="container"><div class="columns"><div class="column three-fourths"><div class="collection-title"><h1 class="collection-header">QIIME2 构建数据库以及常规分析流程</h1><div class="collection-info"> <span class="meta-info"> <span class="octicon octicon-calendar"></span> 2023/10/27 </span> <span class="meta-info"> <span class="octicon octicon-file-directory"></span> <a href="http://localhost:4000/categories/#测序数据分析" title="测序数据分析">测序数据分析</a> </span> <span class="meta-info"> <span class="octicon octicon-clock"></span> 共 6475 字，约 19 分钟 </span></div></div></div><div class="column one-fourth mobile-hidden"><div class="collection-title"></div></div></div></div></section><section class="container content"><div class="columns"><div class="column three-fourths" ><article class="article-content markdown-body"><p>QIIME2 分析流程是扩增子分析常用的流程之一，这里提供一个完整的 16S 数据分析流程，包括构建数据库以及常规分析。QIIME2 是各种插件组成的综合体 pipeline，具体的信息可以看官方网站 <a href="https://docs.qiime2.org/2023.9/">QIIME2</a>.</p><p>构建数据库的过程，参考：<a href="https://forum.qiime2.org/t/processing-filtering-and-evaluating-the-silva-database-and-other-reference-sequence-data-with-rescript/15494">Processing, filtering, and evaluating the SILVA database with RESCRIPt</a></p><h2 id="下载-silva-数据库">下载 SILVA 数据库</h2><p>在 <a href="https://www.arb-silva.de/no_cache/download/archive/release_138.1/Exports/">SILVA</a> 数据库下载并解压如下文件：</p><ul><li>tax_slv_ssu_138.1.txt</li><li>taxmap_slv_ssu_ref_nr_138.1.txt</li><li>tax_slv_ssu_138.1.tre</li><li>SILVA_138.1_SSURef_NR99_tax_silva_trunc.fasta</li></ul><h2 id="初步构建-qiime2-物种分类数据库">初步构建 QIIME2 物种分类数据库</h2><p>Import the Taxonomy Rank file</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime tools import <span class="se">\</span>
    <span class="nt">--type</span> <span class="s1">'FeatureData[SILVATaxonomy]'</span> <span class="se">\</span>
    <span class="nt">--input-path</span> tax_slv_ssu_138.1.txt <span class="se">\</span>
    <span class="nt">--output-path</span> taxranks-silva-138.1-ssu-nr99.qza
</code></pre></div></div><p>Import the Taxonomy Mapping file</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime tools import <span class="se">\</span>
    <span class="nt">--type</span> <span class="s1">'FeatureData[SILVATaxidMap]'</span> <span class="se">\</span>
    <span class="nt">--input-path</span> taxmap_slv_ssu_ref_nr_138.1.txt <span class="se">\</span>
    <span class="nt">--output-path</span> taxmap-silva-138.1-ssu-nr99.qza
</code></pre></div></div><p>Import the Taxonomy Hierarchy Tree file</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime tools import <span class="se">\</span>
    <span class="nt">--type</span> <span class="s1">'Phylogeny[Rooted]'</span> <span class="se">\</span>
    <span class="nt">--input-path</span> tax_slv_ssu_138.1.tre <span class="se">\</span>
    <span class="nt">--output-path</span> taxtree-silva-138.1-nr99.qza
</code></pre></div></div><p>Import the sequence file</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime tools import <span class="se">\</span>
    <span class="nt">--type</span> <span class="s1">'FeatureData[RNASequence]'</span> <span class="se">\</span>
    <span class="nt">--input-path</span> SILVA_138.1_SSURef_NR99_tax_silva_trunc.fasta <span class="se">\</span>
    <span class="nt">--output-path</span> silva-138.1-ssu-nr99-rna-seqs.qza
</code></pre></div></div><p>转换 FeatureData[RNASequence] 为 FeatureData[DNASequence]，获得物种序列</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime rescript reverse-transcribe <span class="se">\</span>
    <span class="nt">--i-rna-sequences</span> silva-138.1-ssu-nr99-rna-seqs.qza <span class="se">\</span>
    <span class="nt">--o-dna-sequences</span> silva-138.1-ssu-nr99-seqs.qza
</code></pre></div></div><p>使用 qiime2 的 silva 分类法创建物种标签分类</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime rescript parse-silva-taxonomy <span class="se">\</span>
    <span class="nt">--i-taxonomy-tree</span> taxtree-silva-138.1-nr99.qza <span class="se">\</span>
    <span class="nt">--i-taxonomy-map</span> taxmap-silva-138.1-ssu-nr99.qza <span class="se">\</span>
    <span class="nt">--i-taxonomy-ranks</span> taxranks-silva-138.1-ssu-nr99.qza <span class="se">\</span>
    <span class="nt">--o-taxonomy</span> silva-138.1-ssu-nr99-tax.qza
</code></pre></div></div><p>——————————————- Optional (可选的) ——————————————-</p><p>也可以直接下载别人构建好的数据库来使用</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime rescript get-silva-data <span class="se">\</span>
    <span class="nt">--p-version</span> <span class="s1">'138.1'</span> <span class="se">\</span>
    <span class="nt">--p-target</span> <span class="s1">'SSURef_NR99'</span> <span class="se">\</span>
    <span class="nt">--o-silva-sequences</span> silva-138.1-ssu-nr99-rna-seqs.qza <span class="se">\</span>
    <span class="nt">--o-silva-taxonomy</span> silva-138.1-ssu-nr99-tax.qza

qiime rescript reverse-transcribe <span class="se">\</span>
    <span class="nt">--i-rna-sequences</span> silva-138.1-ssu-nr99-rna-seqs.qza
    <span class="nt">--o-dna-sequences</span> silva-138.1-ssu-nr99-seqs.qza
</code></pre></div></div><p>——————————————- Optional (可选的) ——————————————-</p><h2 id="对初步数据库进行质控">对初步数据库进行质控</h2><p>删除有 5 个以上模糊碱基的序列，8 个碱基以上的同聚物（Homopolymer 是指基因组上单一碱基重复的区域）</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># 这里用的是默认参数</span>
qiime rescript cull-seqs <span class="se">\</span>
    <span class="nt">--i-sequences</span> silva-138.1-ssu-nr99-seqs.qza <span class="se">\</span>
    <span class="nt">--o-clean-sequences</span> silva-138.1-ssu-nr99-seqs-cleaned.qza
</code></pre></div></div><p>通过序列长度和物种分类信息过滤数据库中的序列，具体地，Archaea (16S) &gt;= 900 bp, Bacteria (16S) &gt;= 1200 bp, and any Eukaryota (18S) &gt;= 1400 bp.</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime rescript filter-seqs-length-by-taxon <span class="se">\</span>
    <span class="nt">--i-sequences</span> silva-138.1-ssu-nr99-seqs-cleaned.qza <span class="se">\</span>
    <span class="nt">--i-taxonomy</span> silva-138.1-ssu-nr99-tax.qza <span class="se">\</span>
    <span class="nt">--p-labels</span> Archaea Bacteria Eukaryota <span class="se">\</span>
    <span class="nt">--p-min-lens</span> 900 1200 1400 <span class="se">\</span>
    <span class="nt">--o-filtered-seqs</span> silva-138.1-ssu-nr99-seqs-filt.qza <span class="se">\</span>
    <span class="nt">--o-discarded-seqs</span> silva-138.1-ssu-nr99-seqs-discard.qza
</code></pre></div></div><p>去除重复冗余的序列</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime rescript dereplicate <span class="se">\</span>
    <span class="nt">--i-sequences</span> silva-138.1-ssu-nr99-seqs-filt.qza  <span class="se">\</span>
    <span class="nt">--i-taxa</span> silva-138.1-ssu-nr99-tax.qza <span class="se">\</span>
    <span class="nt">--p-mode</span> <span class="s1">'uniq'</span> <span class="se">\</span>
    <span class="nt">--o-dereplicated-sequences</span> silva-138.1-ssu-nr99-seqs-derep-uniq.qza <span class="se">\</span>
    <span class="nt">--o-dereplicated-taxa</span> silva-138.1-ssu-nr99-tax-derep-uniq.qza
</code></pre></div></div><h2 id="构建物种分类器">构建物种分类器</h2><p>经过上述处理后，构建 bayes 物种分类器，当然也有其他算法的分类器，可以看 QIIME2 官网，这里的分类器基于的 silva 数据库中 16S 全长序列（部分可能不是）。输出的 silva-138.1-ssu-nr99-classifier.qza 是后续分析可以用到的。</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime feature-classifier fit-classifier-naive-bayes <span class="se">\</span>
  <span class="nt">--i-reference-reads</span>  silva-138.1-ssu-nr99-seqs-derep-uniq.qza <span class="se">\</span>
  <span class="nt">--i-reference-taxonomy</span> silva-138.1-ssu-nr99-tax-derep-uniq.qza <span class="se">\</span>
  <span class="nt">--o-classifier</span> silva-138.1-ssu-nr99-classifier.qza
</code></pre></div></div><p>——————————————- Optional (可选的) ——————————————-</p><p>上述步骤基于的是 16S 全长数据库，可以基于测序区间不同，来构建所需要引物区间的物种分类器，比如 515f-806r</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime feature-classifier extract-reads <span class="se">\</span>
    <span class="nt">--i-sequences</span> silva-138.1-ssu-nr99-seqs-derep-uniq.qza <span class="se">\</span>
    <span class="nt">--p-f-primer</span> GTGYCAGCMGCCGCGGTAA <span class="se">\</span>
    <span class="nt">--p-r-primer</span> GGACTACNVGGGTWTCTAAT <span class="se">\</span>
    <span class="nt">--p-n-jobs</span> 2 <span class="se">\</span>
    <span class="nt">--p-read-orientation</span> <span class="s1">'forward'</span> <span class="se">\</span>
    <span class="nt">--o-reads</span> silva-138.1-ssu-nr99-seqs-515f-806r.qza
</code></pre></div></div><p>在构建所需引物区间后还会再进行去重，固定引物区间后还会有可能有重复的</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime rescript dereplicate <span class="se">\</span>
    <span class="nt">--i-sequences</span> silva-138.1-ssu-nr99-seqs-515f-806r.qza <span class="se">\</span>
    <span class="nt">--i-taxa</span> silva-138.1-ssu-nr99-tax-derep-uniq.qza <span class="se">\</span>
    <span class="nt">--p-mode</span> <span class="s1">'uniq'</span> <span class="se">\</span>
    <span class="nt">--o-dereplicated-sequences</span> silva-138.1-ssu-nr99-seqs-515f-806r-uniq.qza <span class="se">\</span>
    <span class="nt">--o-dereplicated-taxa</span>  silva-138.1-ssu-nr99-tax-515f-806r-derep-uniq.qza
</code></pre></div></div><p>在上述基础上，构建最终的 515f-806r 分类器</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime feature-classifier fit-classifier-naive-bayes <span class="se">\</span>
    <span class="nt">--i-reference-reads</span> silva-138.1-ssu-nr99-seqs-515f-806r-uniq.qza <span class="se">\</span>
    <span class="nt">--i-reference-taxonomy</span> silva-138.1-ssu-nr99-tax-515f-806r-derep-uniq.qza <span class="se">\</span>
    <span class="nt">--o-classifier</span> silva-138.1-ssu-nr99-515f-806r-classifier.qza
</code></pre></div></div><p>——————————————- Optional (可选的) ——————————————-</p><h2 id="数据分析过程">数据分析过程</h2><p>分析前需要表明的是：扩增子数据测序方式和类型有多种，例如 pair-end 双端测序、single-end 单端测序、单一区间测序、16S 全长测序等等，以及测序公司给你返回的原始数据中包不包含引物序列等等，这都需要根据自己的测序数据来具体分析。</p><p>这里提供的分析过程是最常见的一种，双端测序、测序公司去除了引物序列并进行了质控、过程只包括从 fq 数据到 ASV table，物种注释结果以及 ASV 序列。我想后续的分析内容无非是根据自己的实验分组情况，在不同物种分类水平做差异分析。</p><p>制作样本序列信息表 (例如名为 manifest.tsv)，3 列信息分别是样本名称和两端测序数据的路径，形式如下：</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sample-id       forward-absolute-filepath       reverse-absolute-filepath
SRR12466539     /root/pipe_script/qiime2/test/SRR12466539/SRR12466539_1.fastq.gz        /root/pipe_script/qiime2/test/SRR12466539/SRR12466539_2.fastq.gz
SRR12466540     /root/pipe_script/qiime2/test/SRR12466540/SRR12466540_1.fastq.gz        /root/pipe_script/qiime2/test/SRR12466540/SRR12466540_2.fastq.gz
</code></pre></div></div><p>将上一步数据读取</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime tools import <span class="se">\</span>
    <span class="nt">--type</span> <span class="s1">'SampleData[PairedEndSequencesWithQuality]'</span> <span class="se">\</span>
    <span class="nt">--input-path</span> /root/pipe_script/qiime2/test/manifest.tsv <span class="se">\</span>
    <span class="nt">--output-path</span> result/paired-end-demux.qza <span class="se">\</span>
    <span class="nt">--input-format</span> PairedEndFastqManifestPhred33V2
</code></pre></div></div><p>数据质量 summary 可视化</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime demux summarize <span class="se">\</span>
    <span class="nt">--i-data</span> result/paired-end-demux.qza <span class="se">\</span>
    <span class="nt">--o-visualization</span> result/paired-end-demux-summary.qzv
</code></pre></div></div><p>利用 dada2 进行 denoise，我这里使用的是质控好的数据，所以几个质控参数设置的都是 0，即不对序列做处理，也就意味着用 denoise-single 和 denoise-paired 都没问题。</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># --o-representative-sequences 输出的代表序列</span>
<span class="c"># --o-table 输出的 ASV table</span>
<span class="c"># --o-denoising-stats 输出的 denoise 统计信息</span>
qiime dada2 denoise-single <span class="se">\</span>
    <span class="nt">--i-demultiplexed-seqs</span> result/paired-end-demux.qza <span class="se">\</span>
    <span class="nt">--p-trim-left</span> 0 <span class="se">\</span>
    <span class="nt">--p-trunc-len</span> 0 <span class="se">\</span>
    <span class="nt">--p-n-threads</span> 8 <span class="se">\</span>
    <span class="nt">--o-representative-sequences</span> result/rep-seqs-dada2.qza <span class="se">\</span>
    <span class="nt">--o-table</span> result/table-dada2.qza <span class="se">\</span>
    <span class="nt">--o-denoising-stats</span> result/stats-dada2.qza
</code></pre></div></div><p>对 denoise 后的数据进行 summary 可视化</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code>qiime feature-table summarize <span class="se">\</span>
    <span class="nt">--i-table</span> result/table-dada2.qza <span class="se">\</span>
    <span class="nt">--o-visualization</span> result/table-dada2.qzv
</code></pre></div></div><p>利用 bayes 方法做物种注释</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># --o-classification 物种分类结果</span>
<span class="c"># --i-classifier 就是上面构建好的物种分类器</span>
qiime feature-classifier classify-sklearn <span class="se">\</span>
    <span class="nt">--i-classifier</span> /root/database/SILVA_138.1_SSURef_NR99/silva-138.1-ssu-nr99-classifier.qza <span class="se">\</span>
    <span class="nt">--i-reads</span> result/rep-seqs-dada2.qza <span class="se">\</span>
    <span class="nt">--o-classification</span> result/taxonomy-dada2-sliva.qza <span class="se">\</span>
    <span class="nt">--p-n-jobs</span> 3 <span class="se">\</span>
    <span class="nt">--verbose</span>
</code></pre></div></div><p>对上述 qza 结果转换输出为 tsv，fasta 格式文件。</p><div class="language-shell highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="c"># 输出统计信息表</span>
qiime tools <span class="nb">export</span> <span class="se">\</span>
    <span class="nt">--input-path</span>  result/stats-dada2.qza <span class="se">\</span>
    <span class="nt">--output-path</span> result/   <span class="c"># default out file: stats.tsv</span>
<span class="c"># 输出代表序列</span>
qiime tools <span class="nb">export</span> <span class="se">\</span>
    <span class="nt">--input-path</span>  result/rep-seqs-dada2.qza <span class="se">\</span>
    <span class="nt">--output-path</span> result/   <span class="c"># default out file: dna-sequences.fasta</span>
<span class="c"># 输出biom格式的ASV table</span>
qiime tools <span class="nb">export</span> <span class="se">\</span>
    <span class="nt">--input-path</span>  result/table-dada2.qza <span class="se">\</span>
    <span class="nt">--output-path</span> result/   <span class="c"># default out file: feature-table.biom</span>
<span class="c"># 输出物种分类结果</span>
qiime tools <span class="nb">export</span> <span class="se">\</span>
    <span class="nt">--input-path</span>  result/taxonomy-dada2-sliva.qza <span class="se">\</span>
    <span class="nt">--output-path</span> result/  <span class="c"># default out file: taxonomy.tsv</span>
<span class="c"># 转换biom格式为tsv格式</span>
biom convert <span class="nt">-i</span> feature-table.biom <span class="nt">-o</span> feature-table.tsv <span class="nt">--to-tsv</span>
</code></pre></div></div><p>经过上述步骤就得到了扩增子分析中重要的几个结果了，ASV table、ASV 序列和物种注释结果。后续的相关分析就是基于这几个结果去做。</p><div style="margin-top:2em;padding:0 1.5em;border:1px solid #d3d3d3;background-color:#deebf7"><h3>文档信息</h3><ul><li>本文作者：<a href="http://localhost:4000" target="_blank">Wanjin Hu (胡万金)</a></li><li>本文链接：<a href="http://localhost:4000/2023/10/27/pipeline-qiime2-analysis/" target="_blank">http://localhost:4000/2023/10/27/pipeline-qiime2-analysis/</a></li><li>版权声明：自由转载-非商用-非衍生-保持署名（<a href="http://creativecommons.org/licenses/by-nc-nd/3.0/deed.zh" target="_blank">创意共享3.0许可证</a>）</li></ul></div></article><div class="share"></div><div class="comment"> <script src="https://giscus.app/client.js" data-repo="wanjinhu/wanjinhu.github.io" data-repo-id="R_kgDOKJb3gA" data-category="Announcements" data-category-id="DIC_kwDOKJb3gM4CYvQX" data-mapping="title" data-strict="1" data-reactions-enabled="1" data-emit-metadata="0" data-input-position="top" data-theme="light" data-lang="zh-CN" data-loading="lazy" crossorigin="anonymous" async> </script></div></div><div class="column one-fourth"><h3>Search</h3><div id="site_search"> <input style="width:96%" type="text" id="search_box" placeholder="Search"></div><ul id="search_results" style="font-size:14px;list-style-type:none;padding-top:10px;padding-left:10px;"></ul><script src="http://localhost:4000/assets/js/simple-jekyll-search.min.js"></script> <script type="text/javascript"> SimpleJekyllSearch({ searchInput: document.getElementById('search_box'), resultsContainer: document.getElementById('search_results'), json: 'http://localhost:4000/assets/search_data.json?v=1704964685', searchResultTemplate: '<li><a href="{url}" title="{title}">{title}</a></li>', noResultsText: 'No results found', limit: 10, fuzzy: false, exclude: ['Welcome'] }) </script><h3 class="post-directory-title mobile-hidden">Table of Contents</h3><div id="post-directory-module" class="mobile-hidden"><section class="post-directory"><dl></dl></section></div><script src="http://localhost:4000/assets/js/jquery.toc.js"></script></div></div></section><footer class="container"><div class="site-footer" role="contentinfo"><div class="copyright left mobile-block"> © 2023 <span title="Wanjin Hu (胡万金)">Wanjin Hu (胡万金)</span> <a href="javascript:window.scrollTo(0,0)" class="right mobile-visible">TOP</a></div><ul class="site-footer-links right mobile-hidden"><li> <a href="javascript:window.scrollTo(0,0)" >TOP</a></li></ul><a href="https://github.com/wanjinhu/wanjinhu.github.io" target="_blank" aria-label="view source code"> <span class="mega-octicon octicon-mark-github" title="GitHub"></span> </a><ul class="site-footer-links mobile-hidden"><li> <a href="http://localhost:4000/" title="首页" target="">首页</a></li><li> <a href="http://localhost:4000/categories/" title="分类" target="">分类</a></li><li> <a href="http://localhost:4000/archives/" title="归档" target="">归档</a></li><li> <a href="http://localhost:4000/open-source/" title="开源" target="">开源</a></li><li> <a href="http://localhost:4000/wiki/" title="维基" target="">维基</a></li><li> <a href="http://localhost:4000/links/" title="链接" target="">链接</a></li><li> <a href="http://localhost:4000/about/" title="关于" target="">关于</a></li><li><a href="http://localhost:4000/feed.xml"><span class="octicon octicon-rss" style="color:orange;"></span></a></li></ul></div></footer><div class="tools-wrapper"> <a class="gotop" href="#" title="回到顶部"><span class="octicon octicon-arrow-up"></span></a></div><script src="http://localhost:4000/assets/js/geopattern.js"></script> <script> jQuery(document).ready(function($) { $('.geopattern').each(function(){ $(this).geopattern($(this).data('pattern-id')); }); /* hljs.initHighlightingOnLoad(); */ }); </script></body></html>
